File: B트랙/7주차/참고 코드/Fast_Exponentiation.py
import sys, math
input = sys.stdin.readline

def power(n,exp):
    if exp == 0:
        return 1
    mid = power(n, exp // 2)
    return mid * mid if exp % 2 == 0 else mid * mid * n

n,exp = map(int,input().strip().split())
sys.setrecursionlimit(int(math.log2(exp)) + 100)
print(power(n,exp))

File: B트랙/7주차/참고 코드/Fast_Exponentiation.java
import java.io.*;
import java.util.*;

public class Main
{
    public static long power(long n, long exp)
    {
        if(exp == 0) 
        {
            return 1;
        }
        long mid = power(n,exp/2);
        return (exp % 2 == 0) ? mid * mid : mid * mid * n;
    }
    public static void main(String[] args) throws IOException
    {
        BufferedReader scan = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(scan.readLine());
        long n = Long.parseLong(st.nextToken());
        long exp = Long.parseLong(st.nextToken());
        System.out.println(power(n,exp));
        scan.close();
    }
}
